### PR TITLE
First pass at a snippet extractor

### PR DESCRIPTION
**Pull request by jskeet**
Lots more to do, but the results are pretty nice.
Next steps:

- Write a blog post
- Use a script to make all the simplifications instead of adding
  documents to the project
- Revisit all the Roslyn stuff
- Consider separating out into a new project

### CODE REVIEW

File: build/SnippetExtractor/Program.cs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                return 1;
+            }
+        }
+
+        private static async Task ExtractSnippets(string projectFile, string assemblyPaths, string outputDirectory)
+        {
+            // Make sure snippets run in an invariant culture.
+            CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;
+
+            // TODO: Work out why it doesn't work when targeting netcoreapp1.0
```
Comment (malcolmr) [inline build/SnippetExtractor/Program.cs]: Don't you have a github issue for this already?

File: build/SnippetExtractor/Program.cs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+                    var doc = project.GetDocument(entry.Value);
+                    var root = await doc.GetSyntaxRootAsync();
+                    var tree = root.SyntaxTree;
+                    var usings = tree.GetCompilationUnitRoot().Usings.Select(uds => uds.ToString()).ToList();
+                    var method = root.DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
+                    var lines = Trim(method.Body.GetLines());
+
+                    writer.WriteLine("---");
+                    writer.WriteLine($"uid: {uid}");
+                    // TODO: This will overwrite any existing remarks. We'd like a separate "example"
```
Comment (malcolmr) [inline build/SnippetExtractor/Program.cs]: as in, this overwrites `<remarks/>`? That's not great :(

File: build/SnippetExtractor/Program.cs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        }
+
+        private async static Task<Document> RemoveUnusedImportsAsync(Document document)
+        {
+            // TODO: See if there's a better way of doing this.
+            var compilation = await document.Project.GetCompilationAsync();
+            var tree = await document.GetSyntaxTreeAsync();
+            var root = tree.GetRoot();
+            var unusedImportNodes = compilation.GetDiagnostics()
+                .Where(d => d.Id == "CS8019")
```
Comment (malcolmr) [inline build/SnippetExtractor/Program.cs]: Oh, very nice.

File: build/SnippetExtractor/Program.cs (hunk 2 above)
Comment (jskeet) [inline build/SnippetExtractor/Program.cs]: Ah - no, it now doesn't. I fixed the code, but not the TODO. Oops.

I'm expecting to have another iteration within 24 hours, so I'll remove the TODO for that.

File: build/SnippetExtractor/Program.cs (hunk 3 above)
Comment (jskeet) [inline build/SnippetExtractor/Program.cs]: Assuming you were being sarcsastic - I know, it's horrible. I believe there's a nice way of doing this by adding a "simplify" annotation and then asking Roslyn to simplify everything. But I haven't worked it out yet, and I thought I'd give @DustinCampbell a rest after Build...

Comment (malcolmr) [inline build/SnippetExtractor/Program.cs]: Actually, no! I think using the compiler diagnostics is actually a pretty smart way to do this - in the absence of a standard transformation, anyway.

**Pull request merged**